### PR TITLE
Add new capability registry

### PR DESCRIPTION
**Pull request by Revolyssup**
Signed-off-by: Ashish Tiwari <ashishjaitiwari15112000@gmail.com>

**Description**

This PR fixes #

**Notes for Reviewers**

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [ ] Yes, I signed my commits.

### CODE REVIEW

**Comment by Revolyssup** [comment]
The tests will fail until new meshkit release.

**Comment by leecalcote** [comment]
Build failures

File: install/Makefile.core.mk [hunk 1]
```diff
@@ -20,7 +20,7 @@ GIT_COMMITSHA = $(shell git rev-list -1 HEAD)
 GIT_STRIPPED_VERSION=$(shell git describe --tags `git rev-list --tags --max-count=1` | cut -c 2-)
 REMOTE_PROVIDER="remote"
 LOCAL_PROVIDER="local"
-GOVERSION = 1.19.1
+# GOVERSION = 1.19.1
```
Comment (leecalcote) [inline install/Makefile.core.mk:23]: Undo this change, please. Are you not using golang 1.19 on your dev system?

**Review by leecalcote — CHANGES_REQUESTED** [review]
errcheck and staticcheck items....

File: install/docker/Dockerfile [hunk 2]
```diff
@@ -69,6 +69,7 @@ RUN update-ca-certificates
 RUN apk upgrade --no-cache && \
     apk add --no-cache libstdc++
 COPY ./server/oam /app/server/oam
+COPY ./output /app/output
```
Comment (leecalcote) [inline install/docker/Dockerfile:72]: I thought this was to be moved under /server somewhere. Is that right?

File: output/aerokube/2.3.7+2/Config.json [hunk 3]
```diff
@@ -0,0 +1 @@
...},\n    \"endpoint\": {\n     \"default\": \"\",\n     \"type\": \"string\"\n    },\n    \"filename\": {\n     \"default\": \"\",\n     \"type\": \"string\"\n    },\n    \"pattern\": {\n     \"default\": \"\",\n     \"type\": \"string\"\n    },\n    \"secretKey\": {\n     \"default\": \"\",\n     \"type\": \"string\"\n    },\n    \"secretRef\": {\n     \"default\": null,\n     \"properties\": {\n      \"accessKey\": {\n       \"type\": \"string\"\n      },\n      \"name\": {\n       \"pattern\": \"^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$\",\n       \"type\": \"string\"\n      },\n      \"secretKey\": {\n       \"type\": \"string\"\n      }\n     },\n     \"required\": [\n      \"name\"\n     ],\n     \"type\": \"object\"\n    }\n   },\n   \"type\": \"object\"\n  },\n  \"user\": {\n   \"default\": {},\n   \"properties\": {\n    \"id\": {\n     \"default\": 4096,\n     \"format\": \"int64\",\n     \"type\": \"integer\"\n    },\n    \"name\": {\n     \"default\": \"user\",\n     \"pattern\": \"^[a-z_][a-z0-9_-]{0,31}\\\\${0,1}$\",\n     \"type\": \"string\"\n    }\n   },\n   \"type\": \"object\"\n  }\n },\n \"title\": \"Config\",\n \"type\": \"object\"\n}"}
```
Comment (leecalcote) [inline output/aerokube/2.3.7+2/Config.json:1]: Is version 2.3.7+2 really a thing? The "+" seems odd. Does the project use this in their versioning?

Comment (Revolyssup) [reply output/aerokube/2.3.7+2/Config.json:1]: Yes this is what was scraped

File: server/cmd/main.go [hunk 4]
```diff
@@ -38,7 +44,8 @@ var (
 
 const (
 	// DefaultProviderURL is the provider url for the "none" provider
-	DefaultProviderURL = "https://meshery.layer5.io"
+	DefaultProviderURL           = "https://meshery.layer5.io"
+	ArtifactHubComponentsHandler = "kubernetes" //The components generated in output directory will be handled by kubernetes
```
Comment (leecalcote) [inline server/cmd/main.go:48]: Is this same "kubernetes" value being used when the Kubernetes components are registered or is "Meshery-server" used?

Does this variable correspond to host/registrant?

File: server/router/server.go [hunk 5]
```diff
@@ -147,6 +147,10 @@ func NewRouter(ctx context.Context, h models.HandlerInterface, port int, g http.
 	gMux.Handle("/api/meshmodel/validate", h.AuthMiddleware(http.HandlerFunc(h.ValidationHandler), models.NoAuth)).Methods("POST")
 	gMux.Handle("/api/meshmodel/component/generate", h.AuthMiddleware(http.HandlerFunc(h.ComponentGenerationHandler), models.NoAuth)).Methods("POST")
 	gMux.Handle("/api/meshmodel/components", h.AuthMiddleware(http.HandlerFunc(h.GetAllMeshmodelComponents), models.NoAuth)).Methods("GET")
+	gMux.HandleFunc("/api/meshmodel/components/register", h.RegisterMeshmodelComponents).Methods("POST")
+	gMux.HandleFunc("/api/meshmodel/components/types", h.MeshmodelComponentsForTypeHandler).Methods("GET")
+	gMux.HandleFunc("/api/meshmodel/components/{type}", h.GetMeshmodelComponentsByType).Methods("GET")
```
Comment (leecalcote) [inline server/router/server.go:152]: Should these have Auth type assigned?

File: server/cmd/main.go (hunk 4 above)
Comment (Revolyssup) [reply server/cmd/main.go:48]: Yes this is host. Yes this is the same for k8s components as well. Both are handed over to kubernetes.

File: server/router/server.go (hunk 5 above)
Comment (Revolyssup) [reply server/router/server.go:152]: Yes. NoAuth for now. Forgot about it. Making the change now.

**Review by leecalcote — COMMENTED** [review]
This PR feels light on Error Codes.

File: install/Makefile.core.mk (hunk 1 above)
Comment (leecalcote) [reply install/Makefile.core.mk]: Yes/no?

**Comment by Revolyssup** [comment]
> This PR feels light on Error Codes.

This is because most of the heavy lifting code is in meshkit and there I have added few errors. Here one new error is added.

File: install/Makefile.core.mk (hunk 1 above)
Comment (Revolyssup) [reply install/Makefile.core.mk]: Made this change. This was mistake

File: server/cmd/main.go (hunk 4 above)
Comment (leecalcote) [reply server/cmd/main.go:48]: Ok. I thought that we had said `meshery-server` would be registered here, but now I recall how you're short-circuiting with a FK to the `kubernetes-context-id`.

**Comment by leecalcote** [comment]
> > This PR feels light on Error Codes.
> 
> This is because most of the heavy lifting code is in meshkit and there I have added few errors. Here one new error is added.

Oh, I see. Ok.

File: server/handlers/handler_instance.go [hunk 6]
```diff
@@ -36,6 +38,7 @@ func NewHandlerInstance(
 	mctrlHelper *models.MesheryControllersHelper,
 	dbHandler *database.Handler,
```
Comment (leecalcote) [inline server/handlers/handler_instance.go:39]: @Revolyssup regManager works with either sqlite or postgres, right?

File: server/models/meshmodel/core/error.go [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+import (
+	"github.com/layer5io/meshkit/errors"
+)
+
+const (
+	ErrGetK8sComponentsCode = "2212"
+)
+
+func ErrGetK8sComponents(err error) error {
+	return errors.New(ErrGetK8sComponentsCode, errors.Alert, []string{"Could not get K8s components for registration"}, []string{err.Error()}, []string{"Invalid kubeconfig", "Filters passed incorrectly in config", "Could not fetch API resources from Kubernetes server"}, []string{"Make sure that the configuration filters passed are in accordance with output from /openapi/v2"})
```
Comment (leecalcote) [inline server/models/meshmodel/core/error.go:12]: Good work here getting the full set of error code parameters populated.

File: server/models/meshmodel/core/register.go [hunk 8]
```diff
@@ -1,65 +1,282 @@
...
+			Hostname:  "kubernetes",
+			ContextID: contextID,
+		}, v1alpha1.ComponentDefinition{
+			Schema: schema,
+			TypeMeta: v1alpha1.TypeMeta{
+				APIVersion: def.Spec.Metadata["k8sKind"],
+				Kind:       def.Spec.Metadata["k8sAPIVersion"],
+			},
+			Metadata: v1alpha1.ComponentMetadata{
+				Model:   "kubernete",
```
Comment (leecalcote) [inline server/models/meshmodel/core/register.go:68]: "kubernete" on purpose?

File: server/models/meshmodel/core/register.go [hunk 9]
```diff
@@ -1,65 +1,282 @@
...
+	for i, schema := range comp.Schemas {
+		var def oamcore.WorkloadDefinition
+		err := json.Unmarshal([]byte(comp.Definitions[i]), &def)
+		if err != nil {
+			fmt.Println("err here: ", err.Error())
+			return
+		}
+		_ = reg.RegisterEntity(meshmodel.Host{
+			Hostname:  "kubernetes",
+			ContextID: contextID,
```
Comment (leecalcote) [inline server/models/meshmodel/core/register.go:60]: "ContextID" is included in the Meshery Object Model. ✅

File: server/models/meshmodel/core/register.go (hunk 8 above)
Comment (Revolyssup) [reply server/models/meshmodel/core/register.go:68]: oh geez, how do I make such silly mistakes. Thank you for your eyes.

File: server/handlers/handler_instance.go (hunk 6 above)
Comment (Revolyssup) [reply server/handlers/handler_instance.go:39]: Yes. Works with both.

File: server/models/meshmodel/core/register.go (hunk 8 above)
Comment (Revolyssup) [reply server/models/meshmodel/core/register.go:68]: Made the change

**Review by leecalcote — APPROVED** [review]
Good!
